atlasdb-commons/src/main/java/com/palantir/common/concurrent/MultiplexingCompletionService.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.common.concurrent;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RunnableFuture;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A MultiplexingCompletionService is much like a {@link java.util.concurrent.ExecutorCompletionService}, but
+ * supports multiple delegate {@link ExecutorService}s feeding in to a single shared {@link BlockingQueue}.
+ * {@link #poll()} operations will retrieve the results of computations that are done first (regardless of which
+ * actual underlying executor service they may have been scheduled on).
+ *
+ * Maintaining separate executors may see application in improving monitoring and bounding of thread pools that
+ * have several distinct use cases.
+ *
+ * @param <K> key type
+ * @param <V> return type of tasks that are to be submitted
+ */
+public class MultiplexingCompletionService<K, V> {
+    private final ImmutableMap<K, ExecutorService> executors;
+    private final BlockingQueue<Future<V>> taskQueue;
+
+    private MultiplexingCompletionService(
+            ImmutableMap<K, ExecutorService> executors, BlockingQueue<Future<V>> taskQueue) {
+        this.executors = executors;
+        this.taskQueue = taskQueue;
+    }
+
+    public static <T, V> MultiplexingCompletionService<T, V> create(
```

Comment:
nit: if using `<K, V>` then the params here should probably be `<K, V>` too

Reply:
thanks! missed that one